src/StateService/StatePlugin.cs

```diff
@@ -239,5 +242,72 @@ public JObject GetStateHeight(JArray _params)
             json["validatedrootindex"] = StateStore.Singleton.ValidatedRootIndex;
             return json;
         }
+
+        private ContractState GetHistoricalContractState(MPTTrie<StorageKey, StorageItem> trie, UInt160 script_hash)
+        {
+            const byte prefix = 8;
+            StorageKey skey = new KeyBuilder(NativeContract.ContractManagement.Id, prefix).Add(script_hash);
+            return trie[skey]?.GetInteroperable<ContractState>();
+        }
+
+        private void PrepareStateParams(JArray _params, out MPTTrie<StorageKey, StorageItem> trie, out StorageKey skey)
+        {
+            var root_hash = UInt256.Parse(_params[0].AsString());
+            var script_hash = UInt160.Parse(_params[1].AsString());
+            var prefix = Convert.FromBase64String(_params[2].AsString());
+            if (!Settings.Default.FullState && StateStore.Singleton.CurrentLocalRootHash != root_hash)
+                throw new RpcException(-100, "Old state not supported");
+            using var store = StateStore.Singleton.GetStoreSnapshot();
+            trie = new MPTTrie<StorageKey, StorageItem>(store, root_hash);
+
+            var contract = GetHistoricalContractState(trie, script_hash);
+            if (contract is null) throw new RpcException(-100, "Unknown contract");
+            skey = new()
+            {
+                Id = contract.Id,
+                Key = prefix,
+            };
+        }
+
+        [RpcMethod]
+        public JObject FindState(JArray _params)
+        {
+            PrepareStateParams(_params, out var trie, out var skey);
+            StorageKey fkey = new()
+            {
+                Id = skey.Id,
+                Key = Array.Empty<byte>(),
+            };
+            if (3 < _params.Count)
+                fkey.Key = Convert.FromBase64String(_params[3].AsString());
+            int count = Settings.Default.MaxFindResultItems;
+            if (4 < _params.Count)
+                count = int.Parse(_params[4].AsString());
+            if (Settings.Default.MaxFindResultItems < count)
+                count = Settings.Default.MaxFindResultItems;
+            JArray array = new();
+            if (count == 0) return array;
+            int i = 0;
+            foreach (var (ikey, ivalue) in trie.Find(skey.ToArray(), fkey.ToArray()))
```

Comment:
Does you tested with a big tree if we could have a DoS here?

Reply:
I will. But the values are all contract bound. Besides with `from` we only walk through part of tree.

Reply:
> Did you tested with a big tree and result if we could have a DoS here?

100 items costs 5ms, 1000 items costs 25ms from my side. I think it's OK, not very cost time, much better than getNep17transfers.